Onboarding note (#1098)
cahirwpz/mimiker: wiki/docs/onboarding.md

# Onboarding note

## Project overview

> Mimiker is an experimental implementation of very simple UNIX-like operating
> system for Malta board and Raspberry Pi 3.

Reviewer: ```suggestion
Mimiker is an experimental implementation of a simple UNIX-like operating
system for the Malta board and Raspberry Pi 3.
```

system for Malta board and Raspberry Pi 3.

### Main project goals

* to be a research system helping us to understand structure of UNIX kernels
* implement only Kernel of the system on our own.
* keep User Space interface as compatible as possible with
  NetBSD. That way we'll be capable to run all programs designed for that
  system, we already are able to run a few of those. Consequence of that

>   is that all header files are the same as NetBSD's.

Reviewer 2: Which header files are the same? Please specify.

* We want kernel to be run on a modern architecture, we started with MIPS32,
  but currently we are switching to AArch64 (Raspberry Pi 3)
* We love to KISS

### More significant tools we use to run system

> * QEMU - virtual machine allowing us to run system

Reviewer 2: > QEMU - system emulator allowing us to run MIPS and AArch64 guest code on x86-64 hosts.

* Tmux - terminal multiplexer, that allows us to access multiple terminals in
  one window
* Korn Shell (`ksh`) - simple and only one shell that we already ported

#### How system is run?

> When you turn on the system with `./launch` [script][1] system is being run in QEMU.
> Multiple terminals are connected to VM. Between these you can switch with Tmux.
> On one of them is klog (log with kernel messages). On the second one is GDB.
> The other one there can be Korn Shell working in the system.

Reviewer: ```suggestion
When you turn on the system with `./launch` [script][1] system is being run in QEMU.
Multiple terminals are connected to VM. Between these you can switch with Tmux.
One of them is klog (log with kernel messages), the second one is GDB.
The other one there can be Korn Shell working in the system.
```

### Project file hierarchy

Most of the incoherence in file hierarchy come from expectations of C
programs to find header files in specific locations.
More information can be found in [`hier(7)`][2].

* `bin/` and `usr.bin/` programs from NetBSD that are
  considered to be ready to run, because are build during compilation
  * `bin/utest/` tests, that are supposed to be run in User Space,
    testing mostly syscalls
* `contrib/` programs we are capable to build without significant modifications.
  Modifications, if any, are applied as patches
* `build/` build scripts, shared by all Makefiles
* `etc/` system configuration files and scripts
* `include/` folder with all the header files for code ran in User and Kernel
  Space. That one has most of the incoherences, because

>   everything here is compatible with NetBSD and standard C library.

Reviewer: ```suggestion
  everything here is compatible with NetBSD and the standard C library.
```

# Onboarding note

## Project overview

Mimiker is an experimental implementation of a simple UNIX-like operating
system for the Malta board and Raspberry Pi 3.

### Main project goals

* to be a research system helping us to understand structure of UNIX kernels
* implement only Kernel of the system on our own.
* keep User Space interface as compatible as possible with
  NetBSD. That way we'll be capable to run all programs designed for that
  system, we already are able to run a few of those. Consequence of that
  is all user header files are the same as NetBSD's.

> * We want kernel to be run on a modern architecture, we started with MIPS32,
>   but currently we are switching to AArch64 (Raspberry Pi 3)
> * We love to KISS

Reviewer: ```suggestion
* we want kernel to be run on a modern architecture, we started with MIPS32,
  but currently we are switching to AArch64 (Raspberry Pi 3)
* we love to KISS
```
The previous sentences are lower case
